core/jobs/batch_jobs/audit_rule_validation_jobs.py

```diff
@@ -0,0 +1,2194 @@
+            or None if the interaction is not MultipleChoiceInput.
+        """
+        invalid_states = []
+        invalid_states_for_choices_less_than_4 = []
+        for state_name, state in states_dict.items():
+            if state.interaction.id != 'MultipleChoiceInput':
+                continue
+            choices = (
+                state.interaction.customization_args['choices'].value)
+            invalid_choices_index, _ = (
+                ExpAuditRuleChecksJob.
+                _get_invalid_choices_indexes(choices) # type: ignore[arg-type]
+            )
+
+            invalid_ans_groups = []
+            # Remove rules whose choice has been deleted.
+            answer_groups = state.interaction.answer_groups
+            valid_dest_ans_group_count = 0
+            invalid_ans_group_count = 0
+            rule_spec_till_now: List[state_domain.RuleSpecDict] = []
+            for ans_group_idx, answer_group in enumerate(answer_groups):
+                invalid_rules = []
+                for rule_spec in answer_group.rule_specs:
+                    if rule_spec.rule_type == 'Equals':
+                        if rule_spec.inputs['x'] in invalid_choices_index:
+                            invalid_rules.append(rule_spec)
+                    if rule_spec.to_dict() in rule_spec_till_now:
+                        invalid_rules.append(rule_spec)
+                    rule_spec_till_now.append(rule_spec.to_dict())
+
+                if answer_group.outcome.dest != state_name:
+                    valid_dest_ans_group_count += 1
+                    if (
+                        len(invalid_rules) > 0 and
+                        len(set(invalid_rules)) == len(answer_group.rule_specs)
+                    ):
+                        invalid_ans_group_count += 1
+                        invalid_ans_groups.append(ans_group_idx)
+
+            if (
+                invalid_ans_group_count == valid_dest_ans_group_count and
+                invalid_ans_group_count != 0 and
+                valid_dest_ans_group_count != 0
+            ):
+                invalid_states.append(
+                    {
+                        'state_name': state_name,
+                        'ans_group_idx': invalid_ans_groups
+                    }
+                )
+
+            choices_less_than_4 = len(choices) - len(invalid_choices_index) < 4
+            if choices_less_than_4:
+                invalid_states_for_choices_less_than_4.append(state_name)
+
+        return (invalid_states, invalid_states_for_choices_less_than_4)
+
+    @staticmethod
+    def text_input_invalid_values(
+        states_dict: Dict[str, state_domain.State]
+    ) -> List[Dict[str, object]]:
+        """Checks the following in the MultipleChoiceInput interaction
+            - `Contains` should always come after `Equals` and `Startswith`
+            rules
+            - `Startswith` should always come after the `Equals` rule
```

Comment:
These descriptions aren't correct!

Reply:
They're not specific enough -- you can have `Contains hello` before `Equals goodbye`, right? You can have `Startswith hello` before `Equals goodbye`, right?

Reply:
My bad, you are right. Let me edit this.
Thanks